From 628d846c370f94e577d2f1d02ba6aed8ab860c24 Mon Sep 17 00:00:00 2001
From: ywan171 <yi.a.wang@intel.com>
Date: Sun, 12 May 2013 22:33:46 +0800
Subject: [PORT FROM MAIN]libstagefright: avoid ANR in http streaming

BZ: 61400 91117 80682

Orig-patch-id:
89925:Fix anr errors which blocked in setNativeWindow_l when play http straming.
Orig-Change-Id: Idd96ea4f585f0780676f5d49d3c87063c2e940bf

96076:libstagefright: when set new surface, resume the http data fetch after shutdown video Decoder.
Orig-Change-Id: Ic63001d32877fa8a2f7ba51978e903ee910d4ac3

79679:libstagefright: avoid reconnection of http blocking the reset operation
Orig-Change-Id: Ifb258b9e8ebadfb6f44140283b52805f14841823

Change-Id: I18047974a77f1dc4aee43e0065f9557e047b0018
Signed-off-by: ywan171 <yi.a.wang@intel.com>
---
 media/libstagefright/AwesomePlayer.cpp         |   13 +++++++++++++
 media/libstagefright/NuCachedSource2.cpp       |   11 ++++++++++-
 media/libstagefright/include/NuCachedSource2.h |    3 +++
 3 files changed, 26 insertions(+), 1 deletions(-)

diff --git a/media/libstagefright/AwesomePlayer.cpp b/media/libstagefright/AwesomePlayer.cpp
index 984095f..eb1ed63 100644
--- a/media/libstagefright/AwesomePlayer.cpp
+++ b/media/libstagefright/AwesomePlayer.cpp
@@ -499,6 +499,10 @@ void AwesomePlayer::reset_l() {
         addBatteryData(params);
     }
 
+    if (mCachedSource != NULL) {
+        mCachedSource->interrupt(true);
+    }
+
     if (mFlags & PREPARING) {
         modifyFlags(PREPARE_CANCELLED, SET);
         if (mConnectingDataSource != NULL) {
@@ -1258,8 +1262,17 @@ status_t AwesomePlayer::setNativeWindow_l(const sp<ANativeWindow> &native) {
     pause_l();
     mVideoRenderer.clear();
 
+    if (mCachedSource != NULL) {
+        // interrupt the retrying
+        mCachedSource->interrupt(true);
+    }
     shutdownVideoDecoder_l();
 
+    if (mCachedSource != NULL) {
+        // resume the caching
+        mCachedSource->interrupt(false);
+    }
+
     status_t err = initVideoDecoder();
 
     if (err != OK) {
diff --git a/media/libstagefright/NuCachedSource2.cpp b/media/libstagefright/NuCachedSource2.cpp
index 81c9a92..a422c9b 100644
--- a/media/libstagefright/NuCachedSource2.cpp
+++ b/media/libstagefright/NuCachedSource2.cpp
@@ -562,7 +562,7 @@ ssize_t NuCachedSource2::readInternal(off64_t offset, void *data, size_t size) {
 
     size_t delta = offset - mCacheOffset;
 
-    if (mFinalStatus != OK && mNumRetriesLeft == 0) {
+    if (mFinalStatus != OK && (mNumRetriesLeft == 0 || mForceStop)) {
         if (delta >= mCache->totalSize()) {
             return mFinalStatus;
         }
@@ -712,4 +712,13 @@ void NuCachedSource2::RemoveCacheSpecificHeaders(
     }
 }
 
+void NuCachedSource2::interrupt(bool stop) {
+    Mutex::Autolock autolock(mLock);
+    mForceStop = stop;
+    if (mForceStop) {
+        HTTPBase *source = static_cast<HTTPBase *>(mSource.get());
+        source->disconnect();
+    }
+}
+
 }  // namespace android
diff --git a/media/libstagefright/include/NuCachedSource2.h b/media/libstagefright/include/NuCachedSource2.h
index 5db4b4b..a031ed8 100644
--- a/media/libstagefright/include/NuCachedSource2.h
+++ b/media/libstagefright/include/NuCachedSource2.h
@@ -51,6 +51,8 @@ struct NuCachedSource2 : public DataSource {
     size_t cachedSize();
     size_t approxDataRemaining(status_t *finalStatus) const;
 
+    void interrupt(bool stop = true);
+
     void resumeFetchingIfNecessary();
 
     // The following methods are supported only if the
@@ -103,6 +105,7 @@ private:
     off64_t mLastAccessPos;
     sp<AMessage> mAsyncResult;
     bool mFetching;
+    bool mForceStop;
     int64_t mLastFetchTimeUs;
 
     int32_t mNumRetriesLeft;
-- 
1.7.4.1

